cpp/src/io/json/json_tree.cu

```diff
@@ -254,12 +308,423 @@ tree_meta_t get_tree_representation(device_span<PdaTokenT const> tokens,
+                   level,                       \
+                   level_boundaries[level - 1], \
+                   level_boundaries[level],     \
+                   scatter_indices,             \
+                   parent_indices,              \
+                   parent_col_id,               \
+                   node_type,                   \
+                   d_tree.node_levels,          \
+                   col_id);
+#else
+#define PRINT_LEVEL_DATA(level, text) ;
+#endif
+
+  // 4. Propagate parent node ids for each level.
+  // For each level,
+  //     a. gather col_id from previous level results. input=col_id, gather_map is parent_indices.
+  //     b. stable sort by {parent_col_id, node_type}
+  //     c. scan sum of unique {parent_col_id, node_type}
+  //     d. scatter the col_id back to stable node_level order (using scatter_indices)
+  nvtxRangePushA("pre-level");
+  rmm::device_uvector<NodeIndexT> col_id(num_nodes, stream);
+  rmm::device_uvector<NodeIndexT> parent_col_id(num_nodes, stream);
+  thrust::uninitialized_fill(rmm::exec_policy(stream),
+                             parent_col_id.begin(),
+                             parent_col_id.end(),
+                             0);  // XXX: is this needed?
+  // fill with 1, useful for scan later
+  // TODO: Could initialize to 0 and scatter 1 to level_boundaries alone
+  thrust::uninitialized_fill(rmm::exec_policy(stream), col_id.begin(), col_id.end(), 1);
+  // Initialize First level node's node col_id to 0
+  thrust::fill(rmm::exec_policy(stream), col_id.begin(), col_id.begin() + level_boundaries[0], 0);
+  // Initialize First level node's parent_col_id to parent_node_sentinel sentinel
+  thrust::fill(rmm::exec_policy(stream),
+               parent_col_id.begin(),
+               parent_col_id.begin() + level_boundaries[0],
+               parent_node_sentinel);
+  nvtxRangePop();
+  nvtxRangePushA("level");
+  for (decltype(num_levels) level = 1; level < num_levels; level++) {
+    PRINT_LEVEL_DATA(level, ".before gather");
+    nvtxRangePushA("gather");
+    thrust::gather(rmm::exec_policy(stream),
+                   parent_indices.data() + level_boundaries[level - 1],
+                   parent_indices.data() + level_boundaries[level],
+                   col_id.data(),
+                   parent_col_id.data() + level_boundaries[level - 1]);
+    PRINT_LEVEL_DATA(level, ".after gather");
+    nvtxRangePop();
+    nvtxRangePushA("stable_sort_by_key");
+    thrust::stable_sort_by_key(
+      rmm::exec_policy(stream),
+      thrust::make_zip_iterator(parent_col_id.begin() + level_boundaries[level - 1],
+                                node_type.data() + level_boundaries[level - 1]),
+      thrust::make_zip_iterator(parent_col_id.begin() + level_boundaries[level],
+                                node_type.data() + level_boundaries[level]),
+      thrust::make_zip_iterator(scatter_indices.begin() + level_boundaries[level - 1]));
```

Comment:
Since we've seen merge sort taking a considerable share of time in the profiles, we could consider splitting up the sort into two stable sort invocations. Each _stable_ sort to just sort on a single key at a time. 

Would this work?
```suggestion
    // Secondary sort on node_type
    thrust::stable_sort_by_key(
      rmm::exec_policy(stream),
      node_type.data() + level_boundaries[level - 1],
      node_type.data() + level_boundaries[level],
      thrust::make_zip_iterator(
      parent_col_id.begin() + level_boundaries[level - 1], 
      scatter_indices.begin() + level_boundaries[level - 1])
      );
      
    // Primary sort on parent_col_id
    thrust::stable_sort_by_key(
      rmm::exec_policy(stream),
      parent_col_id.data() + level_boundaries[level - 1],
      parent_col_id.data() + level_boundaries[level],
      thrust::make_zip_iterator(
      node_type.begin() + level_boundaries[level - 1], 
      scatter_indices.begin() + level_boundaries[level - 1])
      );
```

Reply:
It works. Wow! 2 radix sorts with 2 values is faster than single merge sort with 2 key and 1 value. 
Radix Sort FTW 🏆

Reply:
ran benchmark in integration PR.
First 3 cases slowed down ~5MB/s. But for large  sizes, ~ 200 MB/s boost!
### [0] Quadro GV100

|    string_size    | Samples |  CPU Time  | Noise  |  GPU Time  | Noise  |  Elem/s  |
|-------------------|---------|------------|--------|------------|--------|----------|
|    2^20 = 1048576 |    976x |   6.159 ms |  4.18% |   6.151 ms |  4.18% | 170.464M |
|    2^21 = 2097152 |   1328x |   6.747 ms |  7.19% |   6.740 ms |  7.19% | 311.136M |
|    2^22 = 4194304 |    992x |   8.718 ms | 18.39% |   8.711 ms | 18.39% | 481.522M |
|    2^23 = 8388608 |    720x |  11.812 ms | 10.28% |  11.805 ms | 10.28% | 710.581M |
|   2^24 = 16777216 |     80x |  20.102 ms |  6.39% |  20.095 ms |  6.39% | 834.903M |
|   2^25 = 33554432 |    469x |  32.013 ms |  6.06% |  32.006 ms |  6.06% |   1.048G |
|   2^26 = 67108864 |     80x |  58.160 ms |  2.49% |  58.153 ms |  2.49% |   1.154G |
|  2^27 = 134217728 |    136x | 110.407 ms |  1.15% | 110.401 ms |  1.15% |   1.216G |
|  2^28 = 268435456 |     72x | 210.488 ms |  0.71% | 210.483 ms |  0.71% |   1.275G |
|  2^29 = 536870912 |     36x | 416.984 ms |  0.78% | 416.981 ms |  0.78% |   1.288G |
| 2^30 = 1073741824 |     11x | 811.059 ms |  0.15% | 811.061 ms |  0.15% |   1.324G |